From 629ad70f43ff1883a8af41ae06f5e0277fc2b6e1 Mon Sep 17 00:00:00 2001
From: Cui GaoSheng <cuigaosheng1@huawei.com>
Date: Thu, 28 Jul 2022 04:52:31 -0400
Subject: [PATCH 3612/4585] [Huawei] efi/libstub: arm64: align the memmap addr
 and reset errant's size

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187251

--------------------------------

Memmap addrs should be aligned by EFI_ALLOC_ALIGN before allocate,
and if allocate_page return failed, reset the size to 0.

Fixes: 3ce00bb63a80 ("[Huawei] efi/libstub: arm64: Fix KASLR and memmap= collision")
Signed-off-by: Cui GaoSheng <cuigaosheng1@huawei.com>
---
 drivers/firmware/efi/libstub/arm64-stub.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/firmware/efi/libstub/arm64-stub.c b/drivers/firmware/efi/libstub/arm64-stub.c
index 4638d511fb1a..143e3c13e742 100644
--- a/drivers/firmware/efi/libstub/arm64-stub.c
+++ b/drivers/firmware/efi/libstub/arm64-stub.c
@@ -73,15 +73,23 @@ void mem_avoid_memmap(void)
 	int i;
 	efi_status_t status;
 	unsigned long nr_pages;
+	unsigned long long start, end;
 
 	for (i = 0; i < MAX_MEMMAP_REGIONS; i++) {
 		if (!mem_avoid[i].size)
 			continue;
-		nr_pages = round_up(mem_avoid[i].size, EFI_ALLOC_ALIGN) / EFI_PAGE_SIZE;
+		start = round_down(mem_avoid[i].start, EFI_ALLOC_ALIGN);
+		end = round_up(mem_avoid[i].start + mem_avoid[i].size, EFI_ALLOC_ALIGN);
+		nr_pages = (end - start) / EFI_PAGE_SIZE;
+
+		mem_avoid[i].start = start;
+		mem_avoid[i].size = end - start;
 		status = efi_bs_call(allocate_pages, EFI_ALLOCATE_ADDRESS,
 				     EFI_LOADER_DATA, nr_pages, &mem_avoid[i].start);
-		if (status != EFI_SUCCESS)
+		if (status != EFI_SUCCESS) {
 			efi_err("Failed to reserve memmap, index: %d, status: %lu\n", i, status);
+			mem_avoid[i].size = 0;
+		}
 	}
 }
 
-- 
2.17.1

